File: erpnext/erpnext/leadership_archival/report/award_master_report/award_master_report.py
# Copyright (c) 2026, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

import frappe


def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    return columns, data

def get_columns():
    return [
        {"label": "Name", "fieldname": "recipientName", "fieldtype": "Data", "width": 150},
        {"label": "CID", "fieldname": "cid", "fieldtype": "Data", "width": 150},
        {"label": "DoB", "fieldname": "dob", "fieldtype": "phone", "width": 150},
        {"label": "Contact No", "fieldname": "contact_no", "fieldtype": "phone", "width": 150},
        {"label": "Position","options": "Position", "fieldname": "position", "fieldtype": "Link", "width": 150},
        {"label": "Organization","options": "Organization", "fieldname": "organization", "fieldtype": "Link", "width": 150},
        {"label": "Confer By","options": "Conferred By", "fieldname": "conferred_by", "fieldtype": "Link", "width": 150},
        {"label": "Start Term","fieldname": "start_term", "fieldtype": "Date", "width": 150},
        {"label": "End Term","fieldname": "end_term", "fieldtype": "Date", "width": 150},
        {"label": "Link to Profile","options": "Key Person Registry", "fieldname": "profile", "fieldtype": "Link", "width": 150},
        {"label": "Kasho","options": "Kasho", "fieldname": "kasho", "fieldtype": "Link", "width": 150},
    ]
    
def get_data(filters):
    conditions = []

    if filters.get("cid"):
        conditions.append("aw.cid = %(cid)s")

    if filters.get("position"):
        conditions.append("aw.position = %(position)s")
    # ✅ FORCE Kasho Type (no frontend dependency)
    conditions.append("k.kasho_type IN ('Appointment')")
        
    start_date = filters.get("start_date")
    end_date = filters.get("end_date")

    if start_date and end_date:
        if end_date < start_date:
            frappe.throw("End Date must be greater than Start Date")
        conditions.append("k.issue_date BETWEEN %(start_date)s AND %(end_date)s")
    elif start_date:
        conditions.append("k.issue_date >= %(start_date)s")
    elif end_date:
        conditions.append("k.issue_date <= %(end_date)s")

    where_clause = " AND ".join(conditions)

    if where_clause:
        where_clause = "WHERE " + where_clause
    else:
        return []  # 🚨 No filters → no data

    return frappe.db.sql(f"""
        SELECT 
			aw.cid AS cid,
			kpr.registry_name AS recipientName,
			kpr.dob AS dob,
			aw.position AS position,
			aw.organization AS organization,
			aw.conferred_by AS conferred_by,
			k.issue_date AS issued_date,
			k.name AS kasho,
			aw.start_term AS start_term,
			aw.end_term as end_term,
			aw.employee_status AS employee_status,
			kpr.name AS profile
		FROM `tabLeadership Appointment` aw
		INNER JOIN `tabKasho` k 
			ON aw.parent = k.name
		LEFT JOIN `tabKey Person Registry` kpr 
			ON kpr.cid = aw.cid
        {where_clause}
    """, filters, as_dict=1)

File: erpnext/erpnext/leadership_archival/report/award_master_report/award_master_report.js
// Copyright (c) 2026, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt

frappe.query_reports["Award Master Report"] = {
	"filters": [
		{
            "fieldname": "cid",
            "label": "CID",
            "fieldtype": "Data",
            "placeholder": "Enter CID"
        },
        {
            "fieldname": "position",
            "label": "Positon",
            "fieldtype": "Link",
            "options": "Position",
            "placeholder": "Select position"
        },
		{
            "fieldname": "start_date",
            "label": "Start Date",
            "fieldtype": "Date",
            "placeholder": "Start Date"
        },
		{
            "fieldname": "end_date",
            "label": "End Date",
            "fieldtype": "Date",
            "placeholder": "End Date"
        },
        {
			"fieldname": "conferred_by",
			"label": "Conferred By",
			"fieldtype": "Link",
			"options": "Conferred By",
			"placeholder": "Select Conferred By"
		},
	],
    // =====================================================
    // FORMATTER
    // =====================================================
    formatter: function (value, row, column, data, default_formatter) {

        value = default_formatter(value, row, column, data);

        if (column.fieldname === "profile" && data && data.profile) {

            return `
                <div style="text-align: center;">
                    <span
                        class="profile-link-icon"
                        data-profile="${data.profile}"
                        title="View Profile"
                        style="cursor: pointer; font-size: 18px;"
                    >
                        <i class="fa fa-eye"></i>
                    </span>
                </div>
            `;
        }

        return value;
    },

        
	onload: function (report) {
        // =====================================================
    // APPLY FILTERS PASSED THROUGH URL
    // =====================================================

    const url_params =
        new URLSearchParams(
            window.location.search
        );

    const position =
        url_params.get("position");

    const conferred_by =
        url_params.get("conferred_by");

    let should_refresh = false;


    if (position) {

        console.log(
            "[Award Master Report] URL Position:",
            position
        );

        report.set_filter_value(
            "position",
            position
        );

        should_refresh = true;
    }


    if (conferred_by) {

        console.log(
            "[Award Master Report] URL Conferred By:",
            conferred_by
        );

        report.set_filter_value(
            "conferred_by",
            conferred_by
        );

        should_refresh = true;
    }


    // =====================================================
    // PROFILE CLICK
    // =====================================================

    report.page.wrapper.on(
        "click",
        ".profile-link-icon",
        function (e) {

            e.stopPropagation();

            const profile_id =
                $(this).data("profile");

            console.log(
                "Profile ID:",
                profile_id
            );

            if (profile_id) {

                const url =
                    `/app/leadership-profile/${profile_id}`;

                window.open(
                    url,
                    "_blank"
                );
            }
        }
    );

        function validate_dates() {
            const start = report.get_filter_value("start_date");
            const end = report.get_filter_value("end_date");

            if (start && end && end < start) {
                frappe.msgprint({
                    title: __("Invalid Date Range"),
                    message: __("End Date must be greater than Start Date"),
                    indicator: "red"
                });

                report.set_filter_value("end_date", null);
                return false;
            }

            return true;
        }

        function refresh_report() {
            if (!validate_dates()) return;

            // ✅ START PROGRESS BAR
            frappe.show_progress(
                __("Loading Report"),
                50,
                100,
                __("Fetching data...")
            );

            report.refresh();

            // ✅ STOP WHEN AJAX FINISHES (ERPNext standard way)
            frappe.after_ajax(() => {
                frappe.show_progress(
                    __("Loading Report"),
                    100,
                    100,
                    __("Done")
                );

                setTimeout(() => {
                    frappe.hide_progress();
                }, 300);
            });
        }

        // Bind filters
        setTimeout(() => {
            const start_filter = report.get_filter("start_date");
            const end_filter = report.get_filter("end_date");

            if (start_filter && end_filter) {
                start_filter.$input.on("change", refresh_report);
                end_filter.$input.on("change", refresh_report);
            }
        }, 500);

        // Initial load
        refresh_report();
    }
};
